test/scale/scale_tester.go

```diff
@@ -117,68 +134,48 @@ scaleOutTest:
 			Replicas: int32(0),
 		},
 	}
+
+	startScaleIn := time.Now()
 	_, _ = client.AppsV1().StatefulSets(statefulSet.Namespace.Name).UpdateScale(context.TODO(), statefulSet.Statefulset.Name, newScale, metav1.UpdateOptions{})
+	numberOfRemainingPods := podutil.WaitForAllPodsWithMatchingLabelToDelete(*scaleTestTimeout, time.Minute, totalNumberOfPods, client, namespace)
+	scaleInPodsCompleted := time.Since(startScaleIn).Minutes()
 
-	numberOfRemainingPods := totalNumberOfPods
-	ticker.Reset(time.Minute)
-	tickerCount = 0
-	timeout = time.After(time.Duration(*scaleTestTimeout) * time.Minute)
-scaleInTest:
-	for {
-		select {
-		case <-timeout:
-			break scaleInTest
-		case <-ticker.C:
-			tickerCount = tickerCount + 1
-			numberOfRemainingPods = podutil.CountAllPodsWithMatchingLabel(client, namespace, map[string]string{"group": "azuredisk-scale-tester"}, string(v1.PodRunning))
-			if numberOfRemainingPods <= 0 {
-				break scaleInTest
-			}
+	numberOfRemainingVolumeAtts := volutil.WaitForAllVolumeAttachmentsToDetach(*scaleTestTimeout, 30*time.Second, totalNumberOfPods**volMountedToPod, client)
+	scaleInVolsDetached := time.Since(startScaleIn).Minutes()
 
-			if tickerCount == 1 {
-				oneMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 2 {
-				twoMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 3 {
-				threeMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 5 {
-				fiveMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 10 {
-				tenMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			}
-		}
+	numberOfRemainingReplicaAtts := 0
+	scaleInAttsCompleted := float64(0)
+	if testWithReplicas {
+		time.Sleep(controller.DefaultTimeUntilGarbageCollection)
```

Comment:
Would it maybe be better to call this sleep in a goroutine after all pods are deleted and then wait for its completion here? That way we will avoid sleeping for more than necessary because I think `WaitForAllVolumeAttachmentsToDetach` might take 2-3 mins or so.

Reply:
Thanks for catching that. I run `WaitForAllVolumeAttachmentsToDetach` in a goroutine instead to avoid it running longer than sleeping time and affecting accuracy of replicas detaching measurement.